src/proxy/resolve.rs

```diff
@@ -129,25 +136,28 @@ where
     R: Resolution,
     R::Endpoint: fmt::Debug,
     R::Error: Into<Error>,
-    M: rt::Make<R::Endpoint>,
+    M: svc::Service<R::Endpoint>,
+    M::Error: Into<Error>,
 {
     type Key = SocketAddr;
-    type Service = M::Value;
+    type Service = M::Response;
     type Error = Error;
 
     fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::Error> {
         loop {
+            if let Async::Ready(Some((addr, svc))) = self.pending.poll().map_err(Into::into)? {
+                self.is_empty.store(false, Ordering::Release);
+                return Ok(Async::Ready(Change::Insert(addr, svc)));
+            }
+
+            try_ready!(self.make.poll_ready().map_err(Into::into));
+
             let up = try_ready!(self.resolution.poll().map_err(Into::into));
             trace!("watch: {:?}", up);
             match up {
                 Update::Add(addr, target) => {
-                    // We expect the load balancer to handle duplicate inserts
-                    // by replacing the old endpoint with the new one, so
-                    // insertions of new endpoints and metadata changes for
-                    // existing ones can be handled in the same way.
-                    let svc = self.make.make(&target);
-                    self.is_empty.store(false, Ordering::Release);
-                    return Ok(Async::Ready(Change::Insert(addr, svc)));
+                    let inner = self.make.call(target);
+                    self.pending.push(MakeFuture { addr, inner });
                 }
                 Update::Remove(addr) => return Ok(Async::Ready(Change::Remove(addr))),
                 Update::NoEndpoints => {
```

Comment:
When we see a `NoEndpoints` update, should we cancel any pending services?

Reply:
This is an extremely good point ;/ I think we'd need something basically like what's described in https://github.com/tower-rs/tower/issues/292... I can probably come up with an approach that doesn't require this futures-unorderd until we have the proper stream data structure...

Reply:
If the balancer always polls discovery until it recieves a NotReady, I don't think the pending services are actually *leaked* --- we should see `Update::Remove` events for them & remove them from the balancer. But, it seems inefficient to drive them to readiness when they'll immediately be torn down...

Reply:
Not really... If we have an Add(A), put it in the pending, Remove(A), and then the pending becomes ready, we'll fire the add out-of-order!

Reply:
oh. yeah, you're right. that's potentially leaky for sure.